module/dkg/broker.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// NewBroker instantiates a new epoch-specific broker capable of communicating
+// with other nodes via a network engine and dkg smart-contract.
+func NewBroker(
+	log zerolog.Logger,
+	dkgInstanceID string,
+	committee flow.IdentifierList,
+	myIndex int,
+	dkgContractClient module.DKGContractClient,
+	tunnel *BrokerTunnel) *Broker {
+
+	b := &Broker{
+		log:               log.With().Str("component", "broker").Str("dkg_instance_id", dkgInstanceID).Logger(),
+		dkgInstanceID:     dkgInstanceID,
+		committee:         committee,
+		myIndex:           myIndex,
+		dkgContractClient: dkgContractClient,
+		tunnel:            tunnel,
+		msgCh:             make(chan messages.DKGMessage),
+	}
+
+	go b.listen()
+
+	return b
+}
+
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Implement DKGBroker
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+// PrivateSend sends a DKGMessage to a destination over a private channel. It
+// appends the current DKG instance ID to the message.
+func (b *Broker) PrivateSend(dest int, data []byte) {
+	if dest >= len(b.committee) || dest < 0 {
+		b.log.Error().Msgf("destination id out of range: %d", dest)
+		return
+	}
+	dkgMessageOut := messages.DKGMessageOut{
+		DKGMessage: messages.NewDKGMessage(b.myIndex, data, b.dkgInstanceID),
+		DestID:     b.committee[dest],
+	}
+	b.tunnel.SendOut(dkgMessageOut)
+}
+
+// Broadcast broadcasts a message to all participants.
+func (b *Broker) Broadcast(data []byte) {
+	dkgMessage := messages.NewDKGMessage(
+		b.myIndex,
+		data,
+		b.dkgInstanceID,
+	)
+	err := b.dkgContractClient.Broadcast(dkgMessage)
+	if err != nil {
+		if err != nil {
+			b.log.Error().
+				Err(err).
+				Msg("could not broadcast message")
+		}
+	}
+}
+
+// Disqualify flags that a node is misbehaving and got disqualified
+func (b *Broker) Disqualify(node int, log string) {}
+
+// FlagMisbehavior warns that a node is misbehaving.
+func (b *Broker) FlagMisbehavior(node int, log string) {}
+
+// GetMsgCh returns the channel through which consumers can receive incoming
+// DKG messages.
+func (b *Broker) GetMsgCh() chan messages.DKGMessage {
+	return b.msgCh
+}
+
+// Poll calls the DKG smart contract to get missing DKG messages for the current
+// epoch, and forwards them to the msgCh. It should be called with the ID of
+// block whose seal is finalized.
+func (b *Broker) Poll(referenceBlock flow.Identifier) error {
+	msgs, err := b.dkgContractClient.ReadBroadcast(b.messageOffset, referenceBlock)
+	if err != nil {
+		return fmt.Errorf("could not read broadcast messages: %w", err)
+	}
+	for _, msg := range msgs {
+		err := b.checkMessageInstanceAndOrigin(msg)
+		if err != nil {
+			b.log.Error().Err(err).Msg("bad broadcast message")
+			continue
+		}
+		b.log.Debug().Msgf("forwarding broadcast message to controller")
+		b.msgCh <- msg
+	}
+	b.messageOffset += uint(len(msgs))
```

Comment:
I'm still worried about concurrent writes here if `Poll` is called concurrently.